fili-core/src/main/java/com/yahoo/bard/webservice/web/responseprocessors/DruidPartialDataResponseProcessor.java

```diff
@@ -0,0 +1,257 @@
+
+    private final ResponseProcessor next;
+
+    /**
+     * Constructor.
+     *
+     * @param next  Next ResponseProcessor in the chain
+     */
+    public DruidPartialDataResponseProcessor(ResponseProcessor next) {
+        this.next = next;
+    }
+
+    @Override
+    public ResponseContext getResponseContext() {
+        return next.getResponseContext();
+    }
+
+    @Override
+    public FailureCallback getFailureCallback(DruidAggregationQuery<?> druidQuery) {
+        return next.getFailureCallback(druidQuery);
+    }
+
+    @Override
+    public HttpErrorCallback getErrorCallback(DruidAggregationQuery<?> druidQuery) {
+        return next.getErrorCallback(druidQuery);
+    }
+
+    /**
+     * If status code is 200, do the following
+     *
+     * <ol>
+     *     <li>
+     *         Extract uncoveredIntervalsOverflowed from X-Druid-Response-Context inside the JsonNode passed into
+     *         DruidPartialDataResponseProcessor::processResponse, if it is true, invoke error response saying limit
+     *         overflowed,
+     *     </li>
+     *     <li>
+     *         Extract uncoveredIntervals from X-Druid-Response-Contex inside the JsonNode passed into
+     *         DruidPartialDataResponseProcessor::processResponse,
+     *     </li>
+     *     <li>
+     *         Parse both the uncoveredIntervals extracted above and allAvailableIntervals extracted from the union of
+     *         all the query's datasource's availabilities from DataSourceMetadataService into SimplifiedIntervalLists,
+     *     </li>
+     *     <li>
+     *         Compare both SimplifiedIntervalLists above, if allAvailableIntervals has any overlap with
+     *         uncoveredIntervals, invoke error response indicating druid is missing some data that are we are expecting
+     *         to exists.
+     *     </li>
+     * </ol>
+     *
+     * @param json  The json representing a druid data response
+     * @param query  The query with the schema for processing this response
+     * @param metadata  The LoggingContext to use
+     */
+    @Override
+    public void processResponse(JsonNode json, DruidAggregationQuery<?> query, LoggingContext metadata) {
+        validateJsonResponse(json, query);
+
+        if (json.get(DruidJsonResponseContentKeys.STATUS_CODE.getName()).asInt() == Status.OK.getStatusCode()) {
+            checkOverflow(json, query);
+
+            SimplifiedIntervalList overlap = getOverlap(json, query);
+            if (!overlap.isEmpty()) {
+                logAndGetErrorCallback(ErrorMessageFormat.DATA_AVAILABILITY_MISMATCH.format(overlap), query);
+            }
+            if (next instanceof FullResponseProcessor) {
+                next.processResponse(json, query, metadata);
+            } else {
+                next.processResponse(json.get(DruidJsonResponseContentKeys.RESPONSE.getName()), query, metadata);
+            }
+        }
+
+        next.processResponse(json, query, metadata);
```

Comment:
If it is 304 and the next response processor is not fullResponseProcessor, we should throw error saying, received 304, but no etag cache response processor is available to process 304 response